pgrepl: parse replication statements + handle IDENTIFY_SYSTEM

otan opened the pull request:
Informs: #105130

## pgwire: ban extended protocol with REPLICATION protocol

In line with PostgreSQL, we prohibit the use of the extended protocol
when the replication protocol is in effect.

Release note: None

## pgrepl: implement IDENTIFY_SYSTEM command

This commit implements IDENTIFY_SYSTEM in the replication protocol,
which is used to retrieve metadata about replication state.

Most of this commit involves changing assumptions throughout the
codebase there is 1 parser - there is 2 if replication mode is enabled.
We've also had to change the parser to return `statements.Statement`
instead of `pgrepltree.ReplicationStatement` for compatibility with
`parser.Parse...`.

I've left the `LSN` fields as placeholders for now as we finalise on
what we'll eventually call it.

Release note: None

## sql: no-op SHOW SYNTAX in replication mode

Currently on the CLI, the command must parse on the server for it to
run. For replication related commands, we have not implemented `SHOW
SYNTAX` yet. This is a bit of effort and seems outside the scope of what
we want for now.

For now, handle replication commands by no-oping them.

Release note: None

otan committed:
pgwire: ban extended protocol with REPLICATION protocol

In line with PostgreSQL, we prohibit the use of the extended protocol
when the replication protocol is in effect.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/106131)
<!-- Reviewable:end -->

otan committed:
pgrepl: implement IDENTIFY_SYSTEM command

This commit implements IDENTIFY_SYSTEM in the replication protocol,
which is used to retrieve metadata about replication state.

Most of this commit involves changing assumptions throughout the
codebase there is 1 parser - there is 2 if replication mode is enabled.
We've also had to change the parser to return `statements.Statement`
instead of `pgrepltree.ReplicationStatement` for compatibility with
`parser.Parse...`.

I've left the `LSN` fields as placeholders for now as we finalise on
what we'll eventually call it.

Release note: None

otan committed:
sql: no-op SHOW SYNTAX in replication mode

Currently on the CLI, the command must parse on the server for it to
run. For replication related commands, we have not implemented `SHOW
SYNTAX` yet. This is a bit of effort and seems outside the scope of what
we want for now.

For now, handle replication commands by no-oping them.

Release note: None

cucaroach reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 7 of 7 files at r1, 21 of 21 files at r2, 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106131#-:-NaquG4VAMCKnQQtvAXB:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rafiss)



<!-- Sent from Reviewable.io -->

otan committed:
pgrepl: use new tests packages

Release note: None

otan commented:
bors r=cucaroach
